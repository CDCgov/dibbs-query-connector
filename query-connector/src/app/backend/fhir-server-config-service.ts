"use server";
import { Pool } from "pg";
import { FhirServerConfig } from "../shared/constants";
import { getDbClient } from "./dbClient";
import { transaction } from "../shared/decorators";

class FhirServerConfigService {
  private dbClient: Pool;
  private cachedFhirServerConfigs: FhirServerConfig[] | null;

  constructor() {
    this.dbClient = getDbClient();
    this.cachedFhirServerConfigs = null;
  }

  /**
   * Fetches the configuration for a FHIR server from the database.
   * @param forceRefresh - Whether to flush the config cache
   * @returns The configuration for the FHIR server.
   */
  async getFhirServerConfigs(forceRefresh = false) {
    if (forceRefresh || this.cachedFhirServerConfigs === null) {
      const query = `SELECT * FROM fhir_servers;`;
      const result = await this.dbClient.query(query);
      const newServerConfigs = result.rows as FhirServerConfig[];
      this.cachedFhirServerConfigs = newServerConfigs;
    }
    return this.cachedFhirServerConfigs;
  }

  async getFhirServerNames(): Promise<string[]> {
    const configs = await this.getFhirServerConfigs();
    return configs.map((config) => config.name);
  }

  /**
   * Updates the connection status for a FHIR server.
   * @param name - The name of the FHIR server
   * @param wasSuccessful - Whether the connection attempt was successful
   * @returns An object indicating success or failure with optional error message
   */
  @transaction
  async updateFhirServerConnectionStatus(name: string, wasSuccessful: boolean) {
    const updateQuery = `
    UPDATE fhir_servers 
    SET 
      last_connection_attempt = CURRENT_TIMESTAMP,
      last_connection_successful = $2
    WHERE name = $1
    RETURNING *;
  `;

    try {
      const result = await this.dbClient.query(updateQuery, [
        name,
        wasSuccessful,
      ]);

      // Clear the cache so the next getFhirServerConfigs call will fetch fresh data
      this.cachedFhirServerConfigs = null;

      if (result.rows.length === 0) {
        return {
          success: false,
          error: "Server not found",
        };
      }

      return {
        success: true,
        server: result.rows[0],
      };
    } catch (error) {
      console.error("Failed to update FHIR server connection status:", error);
      return {
        success: false,
        error: "Failed to update the server connection status.",
      };
    }
  }

  /**
   * Updates an existing FHIR server configuration in the database.
   * @param id - The ID of the FHIR server to update
   * @param name - The new name of the FHIR server
   * @param hostname - The new URL/hostname of the FHIR server
   * @param disableCertValidation - Whether to disable certificate validation
   * @param lastConnectionSuccessful - Optional boolean indicating if the last connection was successful
   * @param bearerToken - Optional bearer token for authentication
   * @returns An object indicating success or failure with optional error message
   */
  @transaction
  async updateFhirServer(
    id: string,
    name: string,
    hostname: string,
    disableCertValidation: boolean,
    lastConnectionSuccessful?: boolean,
    bearerToken?: string,
  ) {
    const updateQuery = `
    UPDATE fhir_servers 
    SET 
      name = $2,
      hostname = $3,
      last_connection_attempt = CURRENT_TIMESTAMP,
      last_connection_successful = $4,
      headers = $5,
      disable_cert_validation = $6
    WHERE id = $1
    RETURNING *;
  `;

    try {
      // If updating with a bearer token, add it to existing headers
      // If no bearer token provided, fetch existing headers and remove Authorization
      let headers = {};
      if (bearerToken) {
        // Get existing headers if any
        const existingServer = await this.dbClient.query(
          "SELECT headers FROM fhir_servers WHERE id = $1",
          [id],
        );
        if (existingServer.rows.length > 0) {
          // Keep existing headers and add/update Authorization
          headers = {
            ...existingServer.rows[0].headers,
            Authorization: `Bearer ${bearerToken}`,
          };
        } else {
          // No existing headers, just set Authorization
          headers = { Authorization: `Bearer ${bearerToken}` };
        }
      } else {
        // Get existing headers if any and remove Authorization
        const existingServer = await this.dbClient.query(
          "SELECT headers FROM fhir_servers WHERE id = $1",
          [id],
        );
        if (existingServer.rows.length > 0) {
          const existingHeaders = existingServer.rows[0].headers || {};
          // Remove Authorization if it exists when switching to no auth
          const { _, ...restHeaders } = existingHeaders;
          headers = restHeaders;
        }
      }

      const result = await this.dbClient.query(updateQuery, [
        id,
        name,
        hostname,
        lastConnectionSuccessful,
        headers,
        disableCertValidation,
      ]);

      // Clear the cache so the next getFhirServerConfigs call will fetch fresh data
      this.cachedFhirServerConfigs = null;

      if (result.rows.length === 0) {
        return {
          success: false,
          error: "Server not found",
        };
      }

      return {
        success: true,
        server: result.rows[0],
      };
    } catch (error) {
      console.error("Failed to update FHIR server:", error);
      return {
        success: false,
        error: "Failed to update the server configuration.",
      };
    }
  }

  /**
   * Inserts a new FHIR server configuration into the database.
   * @param name - The name of the FHIR server
   * @param hostname - The URL/hostname of the FHIR server
   * @param disableCertValidation - Whether to disable certificate validation
   * @param lastConnectionSuccessful - Optional boolean indicating if the last connection was successful
   * @param bearerToken - Optional bearer token for authentication
   * @returns An object indicating success or failure with optional error message
   */
  @transaction
  async insertFhirServer(
    name: string,
    hostname: string,
    disableCertValidation: boolean,
    lastConnectionSuccessful?: boolean,
    bearerToken?: string,
  ) {
    const insertQuery = `
    INSERT INTO fhir_servers (
      name,
      hostname, 
      last_connection_attempt,
      last_connection_successful,
      headers,
      disable_cert_validation
    )
    VALUES ($1, $2, $3, $4, $5, $6);
  `;

    try {
      // Create headers object if bearer token is provided
      const headers = bearerToken
        ? { Authorization: `Bearer ${bearerToken}` }
        : {};

      const result = await this.dbClient.query(insertQuery, [
        name,
        hostname,
        new Date(),
        lastConnectionSuccessful,
        headers,
        disableCertValidation,
      ]);

      // Clear the cache so the next getFhirServerConfigs call will fetch fresh data
      this.cachedFhirServerConfigs = null;

      return {
        success: true,
        server: result.rows[0],
      };
    } catch (error) {
      console.error("Failed to insert FHIR server:", error);
      return {
        success: false,
        error: "Failed to save the FHIR server configuration.",
      };
    }
  }

  /**
   * Deletes a FHIR server configuration from the database.
   * @param id - The ID of the FHIR server to delete
   * @returns An object indicating success or failure with optional error message
   */
  @transaction
  async deleteFhirServer(id: string) {
    const deleteQuery = `
    DELETE FROM fhir_servers 
    WHERE id = $1
    RETURNING *;
  `;

    try {
      const result = await this.dbClient.query(deleteQuery, [id]);

      // Clear the cache so the next getFhirServerConfigs call will fetch fresh data
      this.cachedFhirServerConfigs = null;

      if (result.rows.length === 0) {
        return {
          success: false,
          error: "Server not found",
        };
      }

      return {
        success: true,
        server: result.rows[0],
      };
    } catch (error) {
      console.error("Failed to delete FHIR server:", error);
      return {
        success: false,
        error: "Failed to delete the server configuration.",
      };
    }
  }
}

const configService = new FhirServerConfigService();

export const getFhirServerConfigs =
  configService.getFhirServerConfigs.bind(configService);
export const getFhirServerNames =
  configService.getFhirServerNames.bind(configService);
export const updateFhirServerConnectionStatus =
  configService.updateFhirServerConnectionStatus.bind(configService);
export const updateFhirServer =
  configService.updateFhirServer.bind(configService);
export const insertFhirServer =
  configService.insertFhirServer.bind(configService);
export const deleteFhirServer =
  configService.deleteFhirServer.bind(configService);
