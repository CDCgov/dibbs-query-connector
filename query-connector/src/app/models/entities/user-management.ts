export enum RoleTypeValues {
  Admin = "Admin",
  SuperAdmin = "Super Admin",
  Standard = "Standard User",
}

export interface User {
  id: string;
  username: string;
  first_name: string;
  last_name: string;
  qc_role:
    | RoleTypeValues.SuperAdmin
    | RoleTypeValues.Admin
    | RoleTypeValues.Standard;
  userGroups?: UserGroup[];
  userGroupMemberships?: UserGroupMembership[];
}

export interface UserGroup {
  id: string;
  name: string;
  memberSize: number;
  querySize: number;
}

export interface UserGroupMembership {
  id: string;
  group_name: string;
  usergroup_id: string;
  is_member: boolean;
}

export interface Query {
  id: string;
  name: string;
  userGroupMemberships?: UserGroupMembership[];
}
