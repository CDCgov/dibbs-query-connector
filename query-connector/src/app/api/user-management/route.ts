import { NextResponse } from "next/server";
import { addUserIfNotExists } from "@/app/backend/user-management";

/**
 * API route for adding a user to the user_management table.
 * @param req - The incoming request object.
 * @returns A response indicating success or failure.
 */
export async function POST(req: Request) {
  try {
    const userToken = await req.json();

    if (!userToken?.username) {
      return NextResponse.json(
        { error: "Invalid user token" },
        { status: 400 },
      );
    }

    const user = await addUserIfNotExists(userToken);
    return NextResponse.json(user);
  } catch (error) {
    console.error("Error in user-management API:", error);
    return NextResponse.json(
      { error: "Internal Server Error" },
      { status: 500 },
    );
  }
}
