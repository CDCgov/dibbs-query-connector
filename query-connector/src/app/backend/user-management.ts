"use server";

import { User, UserGroup, UserRole } from "../models/entities/users";
import { QueryTableResult } from "../(pages)/queryBuilding/utils";
import { QCResponse } from "../models/responses/collections";
import { adminAccessCheck, superAdminAccessCheck } from "../utils/auth";
import { getDbClient } from "./dbClient";
const dbClient = getDbClient();

/**
 * Adds a user to the users table if they do not already exist.
 * Uses data extracted from the JWT token.
 * @param userToken - The user data from the JWT token.
 * @param userToken.id - The user ID from the JWT token.
 * @param userToken.username - The username from the JWT token.
 * @param userToken.email - The email from the JWT token.
 * @param userToken.firstName - The first name from the JWT token.
 * @param userToken.lastName - The last name from the JWT token.
 * @param userToken.role - The role from the JWT token.
 * @returns The newly added user or an empty result if already exists.
 */
export async function addUserIfNotExists(userToken: {
  id: string;
  username: string;
  email: string;
  firstName: string;
  lastName: string;
  role?: string;
}) {
  if (!userToken || !userToken.username) {
    console.error("Invalid user token. Cannot add user.");
    return;
  }

  const { username, email, firstName, lastName, role } = userToken;
  const userIdentifier = username || email;

  try {
    console.log("Checking if user exists.");

    const checkUserQuery = `SELECT id, username FROM users WHERE username = $1;`;
    const userExists = await dbClient.query(checkUserQuery, [userIdentifier]);

    if (userExists.rows.length > 0) {
      console.log("User already exists in users:", userExists.rows[0].id);
      return { msg: "User already exists", user: userExists.rows[0] }; // Return existing user
    }

    // Default role when adding a new user, which includes Super Admin, Admin, and Standard User.
    let qc_role = role ?? UserRole.STANDARD;
    console.log("User not found. Proceeding to insert.");

    const insertUserQuery = `
      INSERT INTO users (username, qc_role, first_name, last_name)
      VALUES ($1, $2, $3, $4)
      RETURNING id, username, qc_role, first_name, last_name;
    `;

    const result = await dbClient.query(insertUserQuery, [
      userIdentifier,
      qc_role,
      firstName,
      lastName,
    ]);

    console.log("User added to users:", result.rows[0].id);
    return result.rows[0];
  } catch (error) {
    console.error("Error adding user to users:", error);
    throw error;
  }
}

/**
 * Updates the role of an existing user in the users table.
 * @param userId - The user ID.
 * @param newRole - The new role to assign to the user.
 * @returns The updated user record or an error if the update fails.
 */
export async function updateUserRole(
  userId: string,
  newRole: UserRole,
): Promise<QCResponse<User>> {
  if (!(await superAdminAccessCheck())) {
    throw new Error("Unauthorized");
  }

  if (!userId || !newRole) {
    throw new Error("User ID and new role are required.");
  }

  try {
    console.log(`Updating role for user ID: ${userId} to ${newRole}`);

    const updateQuery = `
      UPDATE users
      SET qc_role = $1
      WHERE id = $2
      RETURNING id, username, qc_role, first_name, last_name;
    `;

    const result = await dbClient.query(updateQuery, [newRole, userId]);

    if (result.rows.length === 0) {
      console.error(`User not found: ${userId}`);
      throw new Error(`User not found: ${userId}`);
    }

    console.log(`User role updated successfully: ${userId} -> ${newRole}`);
    return { totalItems: 1, items: [result.rows[0]] };
  } catch (error) {
    console.error("Error updating user role:", error);
    throw error;
  }
}

/**
 * Retrieves all registered users in query connector
 * @returns List of users registered in qc
 */
export async function getUsers(): Promise<QCResponse<User>> {
  if (!(await adminAccessCheck())) {
    // admins need read access to users for group management
    throw new Error("Unauthorized");
  }

  try {
    const selectAllUsersQuery = `
      SELECT id, username, qc_role, first_name, last_name
      FROM users
      ORDER BY last_name, first_name ASC;
    `;

    const result = await dbClient.query(selectAllUsersQuery);
    const users = await Promise.all(
      result.rows.map(async (user) => {
        try {
          const groups = await getUserGroupsForUser(user.id);
          user.user_groups = groups.items;
          return user;
        } catch (error) {
          console.error(error);
          throw error;
        }
      }),
    );

    return {
      totalItems: result.rowCount,
      items: users,
    } as QCResponse<User>;
  } catch (error) {
    throw error;
  }
}

/**
 * Retrieves a user's role
 * @param username user's username. Username must be unique.
 * @returns The user's role or empty if the user is not found
 */
export async function getUserRole(username: string): Promise<string> {
  try {
    const selectUserRoleQuery = `
      SELECT id, username, qc_role, first_name, last_name
      FROM users
      WHERE username = $1;
    `;

    const result = await dbClient.query(selectUserRoleQuery, [username]);
    return result?.rowCount && result.rowCount > 0
      ? result.rows?.[0].qc_role
      : "";
  } catch (error) {
    throw error;
  }
}

/**
 * Retrieves all registered user groups in query connector along with member and query counts.
 * @returns A list of user groups registered in the query connector.
 */
export async function getUserGroups(): Promise<QCResponse<UserGroup>> {
  if (!(await adminAccessCheck())) {
    throw new Error("Unauthorized");
  }

  try {
    const selectAllUserGroupQuery = `
      SELECT 
        ug.id, 
        ug.name, 
        COALESCE(member_count, 0) AS member_size,
        COALESCE(query_count, 0) AS query_size
      FROM usergroup ug
      LEFT JOIN (
        SELECT usergroup_id, COUNT(*) AS member_count 
        FROM usergroup_to_users 
        GROUP BY usergroup_id
      ) uu ON ug.id = uu.usergroup_id
      LEFT JOIN (
        SELECT usergroup_id, COUNT(*) AS query_count 
        FROM usergroup_to_query 
        GROUP BY usergroup_id
      ) uq ON ug.id = uq.usergroup_id
      ORDER BY ug.name ASC;
    `;

    const result = await dbClient.query(selectAllUserGroupQuery);

    return {
      totalItems: result.rowCount,
      items: result.rows,
    } as QCResponse<UserGroup>;
  } catch (error) {
    console.error("Error retrieving user groups:", error);
    throw error;
  }
}

/**
 * Retrieves the user group(s) for a given user
 * @param id identifier of the us user groups in query connector along with member and query counts.
 * @returns A list of user groups registered in the query connector.
 */
export async function getUserGroupById(id: string): Promise<UserGroup> {
  if (!(await adminAccessCheck())) {
    throw new Error("Unauthorized");
  }

  try {
    const selectAllUserGroupQuery = `
      SELECT id, name 
      FROM usergroup 
      WHERE id = $1 
    `;

    const result = await dbClient.query(selectAllUserGroupQuery, [id]);

    return {
      id: result.rows[0].id,
      name: result.rows[0].name,
      member_size: result.rows[0].member_size,
      query_size: result.rows[0].query_size,
    };
  } catch (error) {
    console.error("Error retrieving user group:", error);
    throw error;
  }
}

/**
 * Retrieves the user group(s) for a given user
 * @param userId identifier of the user whose groups we are retrieving
 * @returns A list of UserGroup items for the given user
 */
export async function getUserGroupsForUser(
  userId: string,
): Promise<QCResponse<UserGroup>> {
  if (!(await adminAccessCheck())) {
    throw new Error("Unauthorized");
  }
  try {
    const selectAllUsersGroupsQuery = `
    SELECT ug.name, ug.id
    FROM usergroup_to_users as ugtu 
    LEFT JOIN users as u ON u.id = ugtu.user_id  
    LEFT JOIN usergroup as ug ON ug.id = ugtu.usergroup_id
    WHERE ugtu.user_id = $1;
    `;

    const result = await dbClient.query(selectAllUsersGroupsQuery, [userId]);
    return {
      totalItems: result.rowCount,
      items: result.rows,
    } as QCResponse<UserGroup>;
  } catch (error) {
    console.error("Error retrieving user groups:", error);
    throw error;
  }
}

/**
 * Creates a new user group if it does not already exist.
 * @param groupName - The name of the user group to create.
 * @returns The created user group or an error message if it already exists.
 */
export async function createUserGroup(
  groupName: string,
): Promise<UserGroup | string> {
  if (!(await superAdminAccessCheck())) {
    throw new Error("Unauthorized");
  }

  try {
    // Check if the group name already exists
    const existingGroups = await getUserGroups();
    const groupExists =
      existingGroups.items?.some((group) => group.name === groupName) ?? false;

    if (groupExists) {
      console.warn(`Group with name '${groupName}' already exists.`);
      return `Group '${groupName}' already exists.`;
    }

    const createGroupQuery = `
      INSERT INTO usergroup (name)
      VALUES ($1)
      RETURNING id, name;
    `;

    const result = await dbClient.query(createGroupQuery, [groupName]);

    return {
      id: result.rows[0].id,
      name: result.rows[0].name,
      member_size: 0,
      query_size: 0,
    };
  } catch (error) {
    console.error("Error creating user group:", error);
    throw error;
  }
}

/**
 * Updates the name of an existing user group.
 * @param id - The unique identifier of the user group to update.
 * @param newName - The new name to assign to the user group.
 * @param userIds - The new name to assign to the user group.
//  * @param queryIds - The new name to assign to the user group.
 * @returns The updated user group or an error if the update fails.
 */
export async function updateUserGroup(
  id: string,
  newName: string,
  userIds?: string[],
  // queryIds?: string[],
): Promise<UserGroup | string> {
  if (!(await superAdminAccessCheck())) {
    throw new Error("Unauthorized");
  }

  try {
    // Check if the new name already exists
    const existingGroups = await getUserGroups();
    const groupExists =
      existingGroups.items?.some((group) => group.name === newName) ?? false;

    if (groupExists) {
      console.warn(`Group with name '${newName}' already exists.`);
      return `Group '${newName}' already exists.`;
    }

    const updateUserGroupMembersQuery = `
      UPDATE usergroup
      SET name = $1
      WHERE id = $2
      RETURNING id, name;
    `;
    const escapedValues =
      userIds && userIds.map((_, i) => `$${i + 1}`).join() + ")";
    const queryString = updateUserGroupMembersQuery + escapedValues;

    const result = await dbClient.query(queryString, [newName, id, userIds]);

    if (result.rows.length === 0) {
      throw new Error(`User group with ID '${id}' not found.`);
    }

    // Get updated memberSize and querySize from getUserGroup
    const userGroups = await getUserGroups();
    const updatedGroup =
      userGroups?.items?.find((group) => group.id === id) ?? null;

    return {
      id: result.rows[0].id,
      name: result.rows[0].name,
      member_size: updatedGroup?.member_size ?? 0,
      query_size: updatedGroup?.query_size ?? 0,
    };
  } catch (error) {
    console.error("Error updating user group:", error);
    throw error;
  }
}

/**
 * Deletes a user group by its unique identifier.
 * @param id - The unique identifier of the user group to delete.
 * @returns The deleted user group or an error if the deletion fails.
 */
export async function deleteUserGroup(id: string): Promise<UserGroup | string> {
  if (!(await superAdminAccessCheck())) {
    throw new Error("Unauthorized");
  }

  try {
    await dbClient.query(
      "DELETE FROM usergroup_to_users WHERE usergroup_id = $1",
      [id],
    );
    await dbClient.query(
      "DELETE FROM usergroup_to_query WHERE usergroup_id = $1",
      [id],
    );
    const deleteGroupQuery = `
      DELETE FROM usergroup
      WHERE id = $1
      RETURNING id, name;
    `;

    const result = await dbClient.query(deleteGroupQuery, [id]);

    if (result.rows.length === 0) {
      throw new Error(`User group with ID '${id}' not found.`);
    }

    return {
      id: result.rows[0].id,
      name: result.rows[0].name,
      member_size: 0,
      query_size: 0,
    };
  } catch (error) {
    console.error("Error deleting user group:", error);
    throw error;
  }
}

/**
 * Retrieves the members for a given user group
 * @param groupId identifier of the group whose members we are retrieving
 * @returns A list of User data for the given group
 */
export async function getGroupMembers(
  groupId: string,
): Promise<QCResponse<User>> {
  try {
    const selectMembersByGroupQuery = `
    SELECT u.id, u.first_name, u.last_name, u.username, u.qc_role
    FROM users as u
    LEFT JOIN usergroup_to_users as ugtu ON ugtu.user_id = u.id
    WHERE ugtu.usergroup_id = $1;
    `;

    const result = await dbClient.query(selectMembersByGroupQuery, [groupId]);
    return {
      totalItems: result.rowCount,
      items: result.rows || [],
    } as QCResponse<User>;
  } catch (error) {
    throw error;
  }
}

/**
 * Retrieves the queries for a given group
 * @param groupId identifier of the group whose queries we are retrieving
 * @returns A list of Query data for the given group
 */
export async function getGroupQueries(
  groupId: string,
): Promise<QCResponse<QueryTableResult>> {
  try {
    const selectQueriesByGroupQuery = `
    SELECT q.id as query_id, q.query_name, q.query_data, q.conditions_list, q.author, q.date_created, q.date_last_modified
    FROM query as q
    LEFT JOIN usergroup_to_query as ugtq ON ugtq.query_id = q.id
    WHERE ugtq.usergroup_id = $1;
    `;
    const result = await dbClient.query(selectQueriesByGroupQuery, [groupId]);
    return {
      totalItems: result.rowCount,
      items: result.rows || [],
    } as QCResponse<QueryTableResult>;
  } catch (error) {
    throw error;
  }
}
