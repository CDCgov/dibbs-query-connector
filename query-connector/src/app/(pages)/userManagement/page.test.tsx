import { ToastContainer } from "react-toastify";
import { render, waitFor, screen } from "@testing-library/react";
import * as UserManagementBackend from "@/app/backend/user-management";
import { RoleTypeValues } from "@/app/models/entities/user-management";
import { RootProviderMock, renderWithUser } from "@/app/tests/unit/setup";
import UserManagement from "./page";

jest.mock("@/app/backend/user-management", () => ({
  getUsers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  updateUserRole: jest.fn(),
}));

describe("User Management: User tab", () => {
  it("Loads user list and not users are found", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UserManagement />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("No users found")).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it("Loads user list successfully", async () => {
    jest.spyOn(UserManagementBackend, "getUsers").mockResolvedValue({
      items: [
        {
          id: "123",
          username: "harrypotter",
          first_name: "harry",
          last_name: "potter",
          qc_role: RoleTypeValues.Admin,
        },
        {
          id: "456",
          username: "lilypotter",
          first_name: "lily",
          last_name: "potter",
          qc_role: RoleTypeValues.Admin,
          userGroups: [
            {
              id: "123",
              name: "Team 1",
              memberSize: 1,
              querySize: 0,
            },
          ],
        },
      ],
      totalItems: 2,
    });

    render(
      <RootProviderMock currentPage="/userManagement">
        <UserManagement />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("potter, harry")).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it("Loads with error while retrieving users", async () => {
    jest
      .spyOn(UserManagementBackend, "getUsers")
      .mockRejectedValueOnce(new Error("Random error"));

    render(
      <RootProviderMock currentPage="/userManagement">
        <UserManagement />
        <ToastContainer />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("Loading...")).toBeInTheDocument();
    expect(screen.getByRole("alert")).toHaveTextContent(
      "Unable to retrieve users. Please try again.",
    );
  });

  it("Changes a user's role", async () => {
    const userMock = {
      id: "123",
      username: "harrypotter",
      first_name: "harry",
      last_name: "potter",
      qc_role: RoleTypeValues.Admin,
    };

    jest.spyOn(UserManagementBackend, "getUsers").mockResolvedValue({
      items: [userMock],
      totalItems: 1,
    });

    const updateRoleFnSpy = jest
      .spyOn(UserManagementBackend, "updateUserRole")
      .mockResolvedValue({
        totalItems: 1,
        items: [
          {
            ...userMock,
            qc_role: RoleTypeValues.SuperAdmin,
          },
        ],
      });

    const { user } = renderWithUser(
      <RootProviderMock currentPage="/userManagement">
        <UserManagement />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    const dropdownEl = screen.getByRole("combobox");
    expect(dropdownEl).toHaveValue("Admin");

    await user.selectOptions(dropdownEl, ["Super Admin"]);

    expect(dropdownEl).toHaveValue("Super Admin");
    expect(updateRoleFnSpy).toHaveBeenCalledWith(userMock.id, "Super Admin");
  });
});
