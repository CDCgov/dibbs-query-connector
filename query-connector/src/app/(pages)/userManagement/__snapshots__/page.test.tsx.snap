// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Management: User tab Loads user list and not users are found 1`] = `
<body>
  <div>
    <div
      class="main-container__wide user-management"
    >
      <h1
        class="page-title"
      >
        User management
      </h1>
      <div
        aria-description="Available user roles"
        class="margin-x-3 margin-y-4 grid-row flex-row flex-no-wrap flex-justify grid-gap"
      >
        <p
          class="grid-col-4"
        >
          <span
            class="text-bold"
          >
            Super Admin:
          </span>
           Manage user permissions; create and manage user groups; view audit logs; configure FHIR servers; create, assign, and run queries
        </p>
        <p
          class="grid-col-4"
        >
          <span
            class="text-bold"
          >
            Admin:
          </span>
           Create, assign, and run queries
        </p>
        <p
          class="grid-col-4"
        >
          <span
            class="text-bold"
          >
            Standard:
          </span>
           Only run queries
        </p>
      </div>
      <div
        class="main-container__wide"
      >
        <div
          class="border-base border-bottom-1px margin-bottom-3"
        >
          <a
            class="tab__active"
            data-prevent-nprogress="true"
            href=""
          >
            Users
          </a>
          <a
            class="tab"
            data-prevent-nprogress="true"
            href=""
          >
            User groups
          </a>
        </div>
        <div
          class="empty-response"
        >
          No users found
        </div>
        <div
          class="drawer closed"
          data-testid="drawer-open-false"
          role="dialog"
        >
          <div
            class="drawerContent display-none"
          >
            <div
              class="drawerHeader"
            >
              <button
                aria-label="Close drawer"
                class="closeButton"
                data-testid="close-drawer"
              >
                <svg
                  aria-label="X icon indicating closure"
                  class="usa-icon usa-icon--size-3"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </button>
              <h2
                class="margin-0 padding-bottom-2"
                data-testid="drawer-title"
              />
              <div>
                <input
                  class="usa-input searchField"
                  data-testid="textInput"
                  id="searchFieldTemplate"
                  name="searchFieldTemplate"
                  placeholder="Search"
                  type="search"
                  value=""
                />
              </div>
            </div>
            <div
              class="padding-top-2"
            >
              <div>
                No queries assigned to this group.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`User Management: User tab Loads user list successfully 1`] = `
<body>
  <div>
    <div
      class="main-container__wide user-management"
    >
      <h1
        class="page-title"
      >
        User management
      </h1>
      <div
        aria-description="Available user roles"
        class="margin-x-3 margin-y-4 grid-row flex-row flex-no-wrap flex-justify grid-gap"
      >
        <p
          class="grid-col-4"
        >
          <span
            class="text-bold"
          >
            Super Admin:
          </span>
           Manage user permissions; create and manage user groups; view audit logs; configure FHIR servers; create, assign, and run queries
        </p>
        <p
          class="grid-col-4"
        >
          <span
            class="text-bold"
          >
            Admin:
          </span>
           Create, assign, and run queries
        </p>
        <p
          class="grid-col-4"
        >
          <span
            class="text-bold"
          >
            Standard:
          </span>
           Only run queries
        </p>
      </div>
      <div
        class="main-container__wide"
      >
        <div
          class="border-base border-bottom-1px margin-bottom-3"
        >
          <a
            class="tab__active"
            data-prevent-nprogress="true"
            href=""
          >
            Users
          </a>
          <a
            class="tab"
            data-prevent-nprogress="true"
            href=""
          >
            User groups
          </a>
        </div>
        <div
          class="customizeTableWrapper"
        >
          <table
            class="usa-table usa-table--borderless customizeTable"
            data-testid="table"
          >
            <thead>
              <tr>
                <th>
                  Name
                </th>
                <th>
                  Permissions
                </th>
                <th>
                  User groups
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  width="270"
                >
                  potter, harry 
          
                </td>
                <td
                  class="roleDropdown"
                  width="270"
                >
                  <label
                    class="usa-label usa-sr-only"
                    data-testid="label"
                    for="role-select-123"
                  >
                    User role
                  </label>
                  <select
                    class="usa-select"
                    data-testid="role-select-123"
                    id="role-select-123"
                  >
                    <option
                      value="Super Admin"
                    >
                      Super Admin
                    </option>
                    <option
                      selected=""
                      value="Admin"
                    >
                      Admin
                    </option>
                    <option
                      value="Standard"
                    >
                      Standard
                    </option>
                  </select>
                </td>
                <td>
                  --
                </td>
              </tr>
              <tr>
                <td
                  width="270"
                >
                  potter, lily 
          
                </td>
                <td
                  class="roleDropdown"
                  width="270"
                >
                  <label
                    class="usa-label usa-sr-only"
                    data-testid="label"
                    for="role-select-456"
                  >
                    User role
                  </label>
                  <select
                    class="usa-select"
                    data-testid="role-select-456"
                    id="role-select-456"
                  >
                    <option
                      selected=""
                      value="Super Admin"
                    >
                      Super Admin
                    </option>
                    <option
                      value="Admin"
                    >
                      Admin
                    </option>
                    <option
                      value="Standard"
                    >
                      Standard
                    </option>
                  </select>
                </td>
                <td>
                  <button
                    aria-description="Edit Team 1 members"
                    class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                    data-testid="button"
                    type="button"
                  >
                    Team 1
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="drawer closed"
          data-testid="drawer-open-false"
          role="dialog"
        >
          <div
            class="drawerContent display-none"
          >
            <div
              class="drawerHeader"
            >
              <button
                aria-label="Close drawer"
                class="closeButton"
                data-testid="close-drawer"
              >
                <svg
                  aria-label="X icon indicating closure"
                  class="usa-icon usa-icon--size-3"
                  focusable="false"
                  height="1em"
                  role="img"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </button>
              <h2
                class="margin-0 padding-bottom-2"
                data-testid="drawer-title"
              />
              <div>
                <input
                  class="usa-input searchField"
                  data-testid="textInput"
                  id="searchFieldTemplate"
                  name="searchFieldTemplate"
                  placeholder="Search"
                  type="search"
                  value=""
                />
              </div>
            </div>
            <div
              class="padding-top-2"
            >
              <div>
                No queries assigned to this group.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
